File: Project3/Problem1.py
cnt = 0

def isSafe(graph, color):

	for i in range(6):
		for j in range(i + 1, 6):
			if (graph[i][j] and color[j] == color[i]):
				return False
	return True

def graphColoring(graph, m, i, color):

	if (i == 6):

		if (isSafe(graph, color)):

			printSolution(color)
		return False

	for j in range(1, m + 1):
		color[i] = j

		if (graphColoring(graph, m, i + 1, color)):
			break
		color[i] = 0
	return False

def printSolution(color):
	global cnt 
	cnt = cnt + 1
	for i in range(6):
		print(color[i], end=" ")
	print()

if __name__ == '__main__':

	graph = [
     	[ 0, 1, 0, 1, 0, 0 ],
		[ 1, 0, 1, 0, 1, 0 ],
		[ 0, 1, 0, 0, 0, 1 ],
		[ 1, 0, 0, 0, 1, 0 ],
		[ 0, 1, 0, 1, 0, 1 ],
		[ 0, 0, 1, 0, 1, 0 ]
  	]
	m = 3

	color = [0 for i in range(6)]
 
	print()

	graphColoring(graph, m, 0, color)
print()
print(cnt)

File: Project3/Problem1.cpp
#include <iostream>
#include <limits>
#include <vector>
#include <time.h>
#include <set>
#include <cstdlib>
#include <random>
#include <stdio.h>

using namespace std;

#define V 6

int cnt = 0;

void printSolution(int color[]);

bool isSafe(bool graph[V][V], int color[])
{
	for (int i = 0; i < V; i++)
		for (int j = i + 1; j < V; j++)
			if (graph[i][j] && color[j] == color[i])
				return false;
	return true;
}

bool graphColoring(bool graph[V][V], int m, int i,
				int color[V])
{
	if (i == V)
	{
		if (isSafe(graph, color))
			printSolution(color);
		return false;
	}

	for (int j = 1; j <= m; j++)
	{
		color[i] = j;

		if (graphColoring(graph, m, i + 1, color))
			break;

		color[i] = 0;
	}

	return false;
}

void printSolution(int color[])
{
	for (int i = 0; i < V; i++)
		cout << " " << color[i];
	cout << "\n";
	cnt++;
}

int main()
{
	bool graph[V][V] = {
		{ 0, 1, 0, 1, 0, 0 },
		{ 1, 0, 1, 0, 1, 0 },
		{ 0, 1, 0, 0, 0, 1 },
		{ 1, 0, 0, 0, 1, 0 },
		{ 0, 1, 0, 1, 0, 1 },
		{ 0, 0, 1, 0, 1, 0 }
	};

	int m = 3;
	
	int color[V];
	for (int i = 0; i < V; i++)
		color[i] = 0;

	if (!graphColoring(graph, m, 0, color))
		// cout << "Solution does not exist";
	
	cout << "Total number of solutions: " << cnt << endl;

	cout << "\n1 corresponds to Red\n";
	cout << "2 corresponds to Blue\n";
	cout << "3 corresponds to Greeen\n";

	return 0;
}